lib/elixir/src/elixir_module.erl

```diff
@@ -176,20 +180,15 @@ build(Line, File, Module, Lexical) ->
   Ref  = elixir_code_server:call({defmodule, self(),
                                  {Module, Data, Defs, Line, File}}),
 
-  DocsOnDefinition =
-      case elixir_compiler:get_opt(docs) of
-        true -> [{'Elixir.Module', compile_doc}];
-        _    -> [{elixir_module, delete_doc}]
-      end,
-
-  ImplOnDefinition =
-      case elixir_config:get(bootstrap) of
-        true -> [{elixir_module, delete_impl}];
-        _    -> [{'Elixir.Module', compile_impl}]
-      end,
+  Bootstrap = elixir_config:get(bootstrap),
+  CompileDocs = elixir_compiler:get_opt(docs),
 
   %% Docs must come first as they read the impl callback.
-  OnDefinition = DocsOnDefinition ++ ImplOnDefinition,
+  OnDefinition = [
+    deprecated_on_definition(not Bootstrap),
+    docs_on_definition(CompileDocs),
+    impl_on_definition(not Bootstrap)
+  ],
```

Comment:
Let's have a general `compile_definition_attributes` function and from there we can call `compile_deprecated` and `compile_impl` but sharing something things, such as the arity and the data table.